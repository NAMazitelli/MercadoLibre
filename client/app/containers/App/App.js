/**
 *
 * App
 * El esqueleto de la pagina
 */

import React from 'react';
import { Helmet } from 'react-helmet';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import SingleProductPage from 'containers/SingleProductPage/Loadable';

import Header from 'components/Header';


const App = () => (
      <div>
        <Helmet
          titleTemplate="%s - MercadoLibre"
          defaultTitle="MercadoLibre"
        >
          <meta name="description" content="MercadoLibre" />
        </Helmet>
        <Header />
        <main>
          <div className="flexbox fullwidth">
            <div className="column is-10-desktop nopad  is-offset-1-desktop">
              <Switch>
                <Route exact path="/" component={HomePage} />
                <Route exact path="/items" component={HomePage} />
                <Route path="/item/:id" component={SingleProductPage} />
                <Route path="" component={NotFoundPage} />
              </Switch>
            </div>
          </div>
        </main>
      </div>
);

export default App;
