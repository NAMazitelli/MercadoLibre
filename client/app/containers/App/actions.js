/*
 * Acciones de la app
 */

import {
  LOAD_PRODUCTS,
  LOAD_PRODUCTS_SUCCESS,
  LOAD_PRODUCTS_ERROR,
  LOAD_SINGLE_PRODUCT,
  LOAD_SINGLE_PRODUCT_SUCCESS,
  LOAD_SINGLE_PRODUCT_ERROR,
  CHANGE_QUERY,
  CHANGE_CATEGORY
} from './constants';

/**
 * Cargar los productos. Esta accion dispara la funcion saga de SearchBox
 *
 * @return {object} un objeto de accion LOAD_PRODUCTS. Si viene un query especifico, con ese query, si no con el que estaba en el estado
 */
export function loadProducts(query = '') {
  if (query !== '') {
    return {
      type: LOAD_PRODUCTS,
      query
    };
  }
  return {
    type: LOAD_PRODUCTS
  };
}
/**
 * Si se cargaron bien los productos, se dispara esta accion.
 *
 * @param  {object} products Los datos de los productos
 *
 * @return {object}      Un objeto de accion con los productos cargados.
 */
export function productsLoaded(products) {
  return {
    type: LOAD_PRODUCTS_SUCCESS,
    products,
  };
}

/**
 * Si se cargaron mal los productos, se dispara esta accion
 *
 * @param  {object} error El error
 *
 * @return {object}       Un objeto de accion con el error.
 */
export function productsLoadingError(error) {
  return {
    type: LOAD_PRODUCTS_ERROR,
    error,
  };
}

/**
 * Cargar un producto por id. Esta accion dispara la funcion saga de SingleProductPage
 *
 * @return {object} un objeto de accion LOAD_SINGLE_PRODUCT con el Id que se pidio
 */
export function loadSingleProduct(queryId) {
  return {
    type: LOAD_SINGLE_PRODUCT,
    queryId
  };
}

/**
 * Si se cargo bien el producto, se dispara esta accion.
 *
 * @param  {object} product Los datos del producto
 *
 * @return {object}      Un objeto de accion con el producto cargado.
 */
export function singleProductLoaded(product) {
  return {
    type: LOAD_SINGLE_PRODUCT_SUCCESS,
    product,
  };
}


/**
 * Si se cargo mal el producto, se dispara esta accion
 *
 * @param  {object} error El error
 *
 * @return {object}       Un objeto de accion con el error.
 */
export function singleProductLoadingError(error) {
  return {
    type: LOAD_SINGLE_PRODUCT_ERROR,
    error,
  };
}
/**
 * Esta accion cambia la query del store por la que le mandamos.
 *
 * @return {object} un objeto de accion CHANGE_QUERY con el query ingresado
 */
export function changeQuery(query) {
  return {
    type: CHANGE_QUERY,
    query
  };
}
/**
 * Esta accion cambia la categoria del store por la que le mandamos. Usado para el breadcrumbs.
 *
 * @return {object} un objeto de accion CHANGE_CATEGORY con la categoria ingresada
 */
export function changeCategory(category) {
  return {
    type: CHANGE_CATEGORY,
    category
  };
}
