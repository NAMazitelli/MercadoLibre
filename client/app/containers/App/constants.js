export const LOAD_PRODUCTS = 'MercadoLibre/App/LOAD_PRODUCTS';
export const LOAD_PRODUCTS_SUCCESS = 'MercadoLibre/App/LOAD_PRODUCTS_SUCCESS';
export const LOAD_PRODUCTS_ERROR = 'MercadoLibre/App/LOAD_PRODUCTS_ERROR';

export const LOAD_SINGLE_PRODUCT = 'MercadoLibre/App/LOAD_SINGLE_PRODUCT';
export const LOAD_SINGLE_PRODUCT_SUCCESS = 'MercadoLibre/App/LOAD_SINGLE_PRODUCT_SUCCESS';
export const LOAD_SINGLE_PRODUCT_ERROR = 'MercadoLibre/App/LOAD_SINGLE_PRODUCT_ERROR';

export const CHANGE_QUERY = 'MercadoLibre/App/CHANGE_QUERY';
export const CHANGE_CATEGORY = 'MercadoLibre/App/CHANGE_CATEGORY';
