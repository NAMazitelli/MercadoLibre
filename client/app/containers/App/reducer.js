import {
  LOAD_PRODUCTS,
  LOAD_PRODUCTS_SUCCESS,
  LOAD_PRODUCTS_ERROR,
  LOAD_SINGLE_PRODUCT,
  LOAD_SINGLE_PRODUCT_SUCCESS,
  LOAD_SINGLE_PRODUCT_ERROR,
  CHANGE_QUERY
} from './constants';

// El estado inicial
export const initialState = {
  loading: false,
  error: false,
  productData: {
    list: false,
    product: false,
  },
  category: '',
  query: '',
  queryId: ''
};

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_PRODUCTS: {
      if (action.query !== undefined) {
        return {
          ...state,
          loading: true,
          error: false,
          productData: {
            list: false,
            product: false,
          },
          query: action.query,
        };
      }
      return {
        ...state,
        loading: true,
        error: false,
        productData: {
          list: false,
          product: false,
        }
      };
    }

    case LOAD_PRODUCTS_SUCCESS: {
      const newState = {
        ...state,
        loading: false,
        productData: {
          list: action.products,
          product: false,
        },
        category: action.products.categories
      };
      return newState;
    }

    case LOAD_PRODUCTS_ERROR: {
      return { ...state, error: action.error, loading: false };
    }

    case LOAD_SINGLE_PRODUCT: {
      const newState = {
        ...state,
        loading: true,
        error: false,
        productData: {
          list: false,
          product: false
        },
        queryId: action.queryId
      };

      return newState;
    }

    case LOAD_SINGLE_PRODUCT_SUCCESS: {
      const newState = {
        ...state,
        loading: false,
        productData: {
          list: false,
          product: action.product,
        },
        category: action.product.item.categories
      };

      return newState;
    }

    case LOAD_SINGLE_PRODUCT_ERROR: {
      return { ...state, error: action.error, loading: false };
    }

    case CHANGE_QUERY:
        return { ...state, query: action.query };


    default:
      return state;
  }
}

export default appReducer;
