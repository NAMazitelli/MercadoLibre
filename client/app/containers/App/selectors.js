import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = (state) => state.global || initialState;

const makeSelectLoading = () => createSelector(
  selectGlobal,
  (globalState) => globalState.loading
);

const makeSelectError = () => createSelector(
  selectGlobal,
  (globalState) => globalState.error
);

const makeSelectProductsList = () => createSelector(
  selectGlobal,
  (globalState) => globalState.productData.list
);

const makeSelectProduct = () => createSelector(
  selectGlobal,
  (globalState) => globalState.productData.product
);

const makeSelectQueryId = () => createSelector(
  selectGlobal,
  (globalState) => globalState.queryId
);

const makeSelectQuery = () => createSelector(
  selectGlobal,
  (globalState) => globalState.query
);

const makeSelectCategory = () => createSelector(
  selectGlobal,
  (globalState) => globalState.category
);

export {
  selectGlobal,
  makeSelectLoading,
  makeSelectError,
  makeSelectProductsList,
  makeSelectProduct,
  makeSelectQueryId,
  makeSelectQuery,
  makeSelectCategory
};
