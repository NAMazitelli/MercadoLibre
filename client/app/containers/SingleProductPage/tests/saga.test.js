/**
 * Testeo de sagas de SingleProductPage
 */

import { put, takeLatest } from 'redux-saga/effects';
import { LOAD_SINGLE_PRODUCT } from 'containers/App/constants';
import { singleProductLoaded, singleProductLoadingError } from 'containers/App/actions';
import singleProductData, { getSingleProduct } from '../saga';

const id = 'MLA733099149';

/* eslint-disable redux-saga/yield-effects */
describe('getSingleProduct Saga', () => {
  let getSingleProductGenerator;

  beforeEach(() => {
    getSingleProductGenerator = getSingleProduct();

    const selectDescriptor = getSingleProductGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getSingleProductGenerator.next(id).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('Deberia disparar la accion de singleProductLoaded si se puede traer la data', () => {
    const response = [{
      name: 'First item',
    }, {
      name: 'Second item',
    }];
    const putDescriptor = getSingleProductGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(singleProductLoaded(response, id)));
  });

  it('Deberia disparar la accion singleProductLoadingError si hubo errores', () => {
    const response = new Error('Some error');
    const putDescriptor = getSingleProductGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(singleProductLoadingError(response)));
  });
});

describe('singleProductData Saga', () => {
  const singleProductDataSaga = singleProductData();

  it('Deberia esperar acciones de LOAD_SINGLE_PRODUCT', () => {
    const takeLatestDescriptor = singleProductDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_SINGLE_PRODUCT, getSingleProduct));
  });
});
