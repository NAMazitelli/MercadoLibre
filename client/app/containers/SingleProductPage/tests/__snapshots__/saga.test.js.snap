// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSingleProduct Saga Deberia disparar la accion de singleProductLoaded si se puede traer la data 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getSingleProduct Saga Deberia disparar la accion de singleProductLoaded si se puede traer la data 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:9000/api/items/MLA733099149",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getSingleProduct Saga Deberia disparar la accion singleProductLoadingError si hubo errores 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getSingleProduct Saga Deberia disparar la accion singleProductLoadingError si hubo errores 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:9000/api/items/MLA733099149",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
